src/main/ruby/core/posix.rb

```diff
@@ -7,20 +7,42 @@
 # GNU Lesser General Public License version 2.1
 
 module Truffle::POSIX
-  LIBC = Truffle::Interop.eval('application/x-native', 'default')
+  class LazyLibrary
+    def initialize(&block)
+      @block = block
+      @resolved = nil
+    end
 
-  home = Truffle::Boot.ruby_home
-  libtruffleposix = "#{home}/lib/cext/truffleposix.#{Truffle::Platform::NATIVE_DLEXT}" if home
-  if home
-    LIBTRUFFLEPOSIX = Truffle::Interop.eval('application/x-native', "load '#{libtruffleposix}'")
-  else
-    LIBTRUFFLEPOSIX = LIBC
+    def resolve
+      if resolved = @resolved
+        resolved
+      else
+        Truffle.synchronize(self) do
```

Comment:
Couldn't you make this a one-liner with `@resolved ||= Truffle.synchronize(self) { @block.call }` ?

Reply:
That could call the block twice.